### ISSUE TITLE
Roslyn doesn't report warnings for interator methods with [DoesNotReturn] annotation

### ISSUE THREAD

**Issue by TessenR** [dotnet/roslyn#43228]
**Version Used**: 

```
commit ff3d2a75dcd9ad286d4bad47c993454f3d1dd1d7
Merge: 7be257abe9 6fd1f755aa
Author: msftbot[bot] <48340428+msftbot[bot]@users.noreply.github.com>
Date:   Thu Apr 2 15:49:03 2020 +0000

    Merge pull request #43000 from mavasani/PortConvertAnonymousTypeToTuple

    Port ConvertAnonymousTypeToTuple to shared layer
```

**Steps to Reproduce**:

```cs
#nullable enable
using System.Collections.Generic;
using System.Diagnostics.CodeAnalysis;
class C
{
    [DoesNotReturn] public static IEnumerable<int> M() { yield break; }
    
    public static void Main()
    {
      string? s = null;
      if (s == null)
          M();
        
      s.ToString();
    }
}
```

**Expected Behavior**:
`warning CS8763: A method marked [DoesNotReturn] should not return.` reported for `M`

**Actual Behavior**:
No warnings. The program crashes at runtime with a `NullReferenceException`
